exasol_advanced_analytics_framework/udf_communication/peer_communicator/background_listener_thread.py

```diff
@@ -146,31 +150,51 @@ def _create_poller(self):
 
     def _run_message_loop(self):
         try:
-            while self._status == BackgroundListenerThread.Status.RUNNING:
-                poll = self.poller.poll(timeout_in_ms=self._config.poll_timeout_in_ms)
-                if self._in_control_socket in poll and PollerFlag.POLLIN in poll[self._in_control_socket]:
-                    message = self._in_control_socket.receive()
-                    self._status = self._handle_control_message(message)
-                if self._listener_socket in poll and PollerFlag.POLLIN in poll[self._listener_socket]:
-                    message = self._listener_socket.receive_multipart()
-                    self._handle_listener_message(message)
-                if self._status == BackgroundListenerThread.Status.RUNNING:
-                    for peer_state in self._peer_state.values():
-                        peer_state.resend_if_necessary()
+            while self._status != BackgroundListenerThread.Status.STOPPED:
+                self._handle_message()
+                self._try_send()
+                self._check_is_ready_to_stop()
         except Exception as e:
             self._logger.exception("Exception in message loop")
 
+    def _check_is_ready_to_stop(self):
+        if self._status == BackgroundListenerThread.Status.PREPARE_TO_STOP:
+            if self._is_ready_to_stop():
+                self._out_control_socket.send(serialize_message(IsReadyToStop()))
+
+    def _is_ready_to_stop(self):
+        peers_status = [peer_state.is_ready_to_stop()
+                        for peer_state in self._peer_state.values()]
+        is_ready_to_stop = all(peers_status) and len(peers_status) == self._number_of_peers - 1
```

Comment:
How  could ` len(peers_status) != self._number_of_peers - 1` happen ? Consequences would be that it never reaches ready_to_stop, right?

Reply:
This is a common misunderstanding. Each peer stores its state in relation to each other peer. That means we only have a peer_state for the other peers, but not for us self. With that, we only have self._number_of_peers-1 many peers_states.

Reply:
My question is more related to the case when a peer might be down or has nw issues or whatever, then the background_listener_thread never reach ready_to_stop. If this is intended?

Reply:
Ah, now I understand the question. Yes, this is intended. When this happens, the stop timeout sets in and stop everything never the less, but ends with an exception. The reason, for this behavior is, if we can't successful stop everything it is likely, that one of the other peers is not reachable or stuck, in this case we can't successful end, but we don't want to wait forever either.